Enable editing of imported resources when editing snippets

piemonkey opened the pull request:
## Overview
Move to editor and plugins that support rdfa linking of imported resources in snippets. Biggest change here is to update the list of imported resources on the snippet list whenever saving a snippet.

Known issues:
- [x] Deleting a relationship on a document node (so an imported resource relationship) doesn't work, it's necessary to delete the backlink and reload
- [ ] Relationship editor doesn't display the subject for imported resource relationships, which is confusing
- [ ] There's no good handling of unlinked imported resources, e.g. preventing insertion if not linked or just removing the backlinks
- [x] Deleting snippets from a snippet list does not update the list of imported resources for that list (e.g. if an import is no longer used by any snippets, it should be removed but is not).

##### connected issues and PRs:
Plugins: https://github.com/lblod/ember-rdfa-editor-lblod-plugins/pull/462
Editor: https://github.com/lblod/ember-rdfa-editor/pull/1212
backend: https://github.com/lblod/app-reglementaire-bijlage/pull/81
For Jira ticket [GN-4898](https://binnenland.atlassian.net/browse/GN-4898)
GN: https://github.com/lblod/frontend-gelinkt-notuleren/pull/704

### Setup
Needs to be run with a backend that supports storing imported resources on snippet lists.

### How to test/reproduce
There are 3 steps to this:
1. Create a snippet with imported resources:
	- Go to the snippets page
	- Create a new or edit an existing snippet list, then edit or create a snippet
	- Create arbitrary RDFa in the editor
	- Click on the 'document' node (so click on the page outside of the content), and verify that the RDFa tools give the type 'document'
	- Open the 'Add Relationship' modal, create a link as normal, except also specify the subject. For the first imported resource you'll need to enter the placeholder URI into the power-select 'search' and hit enter, this will set the field to this new value. Later you can select an existing import or create a new one in the same way.
	- Hit save
	- Go back to 'edit snippet list' or 'manage snippet lists' pages and verify that the snippet lists the union of all the imported resources used in each of the snippets in that list
2. Create a template containing a snippet placeholder:
	- Go to the template management page and create or edit a template
	- Create some arbitrary RDFa, including at least one resource
	- Insert a snippet placeholder which incudes imported resources
	- Select the placeholder and see that the RDFa relationship editor has a new section listing the imported resources. You can click on the `+` to choose the 'real' resource URI that it links to in the document. Clicking the name of the linked 'real' resource allows you to change an already configured link.
	- Click save and publish, then publish.
3. Create a real document from a template with a snippet placeholder:
	- See [PR in GN](https://github.com/lblod/frontend-gelinkt-notuleren/pull/704)

### Challenges/uncertainties
N/A

### Checks PR readiness
- [ ] UI: works on smaller screen sizes
- [ ] UI: feedback for any loading/error states
- [ ] Check cancel/go-back flows
- [ ] Check database state correct when deleting/updating (especially regarding relationships)
- [x] changelog
- [x] npm lint
- [x] no new deprecations

elpoelma reviewed (commented):
Tested using a simple snippet with two literal nodes connect to 1 imported resource. Will do more extensive testing later on.

At first sight, seems to work really nicely. Tested by creating a snippet, adding the snippet list to a placeholder in a template, connecting the imported resources, and creating an AP in GN based on the template. 

I did notice something that might be an issue:
When having a snippet list with 2 snippets and creating an imported resource in 1 of the 2 snippets, the imported resource does not become available in the dropdown of the other one (you have to manually type it). I assume this is due to it not being added to the data-attribute.

Will leave a more extensive review tomorrow.

abeforgit commented:
_this is so cool_
absolutely incredible work

elpoelma reviewed (changes requested):
Overall, seems to work really really well. Excited for the future :)
Creating snippet-lists with different snippets connecting to the same imported resources works very well. Configuring the imported resources of the snippet lists in the placeholder is quite intuitive.
I did encounter some issues:
- When having a snippet-list with 2 snippets, and using the swap functionality on the GN frontend, the imported resources of the swapped snippet do not seem to connect correctly => backlinks are added, properties not. Note: this could be related to a bug I solved on the editor but is not yet merged (https://github.com/lblod/ember-rdfa-editor/pull/1211), not sure
- As mentioned before, imported resources created in one snippet are not automatically suggested in another. We could probably solve this by two-way syncing the imported-resources: snippet-list <-> snippet.

Otherwise, seems to really work well! Still have to test nested snippets.

piemonkey commented:
@elpoelma I've added logic [to the plugins](https://github.com/lblod/ember-rdfa-editor-lblod-plugins/pull/462/files#diff-3a45d2166eea85bf7ba90956072d3a3ae7461dd22724561246fb2b3eac0ada81) to set up the linking correctly with the outside document when inserting a snippet. I haven't tested this extensively yet, but at least fixes the issue you noticed.

I also added the ability to pass imported resources into the doc from outside, and use this for the other resources in the snippet list, which makes it much more sane to work on multiple snippets expecting the same resources.

piemonkey commented:
FYI, I've added a list of known issues to the description, so far none would stop this from being merged if we just want to fix them in later PRs.

elpoelma reviewed (commented):
Did another test-run in the snippet editor, did encounter some issues:
- Removing snippets do not cause imported resources to go away (not a problem perse)
- Connecting literals to previously defined imported resources does not work:
https://github.com/user-attachments/assets/aade7fc0-70b6-4846-a510-5eb5b4869db1

elpoelma commented:
> You mean the problem noted in the description where deleting a relationship on the document node doesn't work? Or a different problem?

No, I mean that removing a whole snippet document from a snippet list altogether, doesn't cause the referenced (let's call them dangling/unused) imported resources to disappear from the snippet list.



> This is actually a wider problem. Adding a relationship between any resource and a literal with an existing backlink does not work. While it may work in the RDFa tools, the HTML produced does not actually contain that link. I discussed this with @abeforgit and wondered whether it conceptually makes sense to have a literal with more than one backlink. I decided that it wasn't something to fix in this work. (by fix, I mean either make it supported or remove the possibility to create these links).

It's not exactly this problem I'm referencing. If you look at the video, the (newly created) literal to which a relationship is added has no backlinks yet. The problem is re-using an imported resource defined in another snippet.

piemonkey commented:
> No, I mean that removing a whole snippet document from a snippet list altogether, doesn't cause the referenced (let's call them dangling/unused) imported resources to disappear from the snippet list.

Ah, I've added this to the list of issues to fix.

> It's not exactly this problem I'm referencing. If you look at the video, the (newly created) literal to which a relationship is added has no backlinks yet. The problem is re-using an imported resource defined in another snippet.

Oops, I missed the video. Can you confirm if you're seeing this with the latest `editor`? I spotted a problem with actually using any of the resources imported in other documents just after putting that in and [added a fix](https://github.com/lblod/ember-rdfa-editor/pull/1212/commits/6b2952177ee4b12164e36807056de74e447b17fd).

abeforgit commented on app/components/snippet-list-form.js:
```diff
@@ -4,6 +4,7 @@ import { action } from '@ember/object';
 import { restartableTask, task, timeout } from 'ember-concurrency';
 import { tracked } from '@glimmer/tracking';
 import { isBlank } from '../utils/strings';
+import { saveCollatedImportedResources } from '../utils/imported-resources';
```
seems like this util got (re)moved? it's throwing an error that it can't find the import, and neither can I by grepping in the sourcecode

piemonkey replied:
Damn. Fixed that now, sorry about that.

elpoelma reviewed (approved):
Tested again, it works fantastically. The breaking issues all seem to be solved.
- Removing snippets from a snippet-list leads to a recalculation of the imported-resources
- Replace a snippet with another one works well
- Removing snippets works well
- Re-using already defined imported resources works well

Really cool!
